tests/integration/framework/docker_runner.py

```diff
@@ -8,49 +8,93 @@
 import sys
 
 class DockerRunner:
-    """Run docker containers for testing"""
+    """Run a docker container in background to execute testing commands"""
 
     def __init__(self, platform, image, verbose):
         """Sets platform and image for all tests ran with this instance"""
         self.platform = platform
         self.image = image
         self.verbose = verbose
+        self.container_id = None
 
-    def construct_docker_command(self, envs, args):
+        # Launch the container
+        self._start()
+
+    def execute(self, envs, args):
+        """Launch `docker exec` commands inside the background container"""
+        command = self._construct_command("exec", envs, args)
+        return self._shell(command)
+
+    def run(self, envs, args):
+        """
+        Launch `docker run` commands to create a new container for each command
+        """
+        command = self._construct_command("run", envs, args)
+        return self._shell(command)
+
+    def __del__(self):
+        """Remove test container"""
+        stop_command = ["docker", "rm", "-f", self.container_id]
+        self._shell(stop_command, silent=True)
+
+    def _start(self):
```

Comment:
Why protected?

Reply:
Just a habit because it wasn't used by outside. It was designed with the idea that DockerRunner will always start a container for a test.

Reply:
I'm okay with it when it's designed to not be ran from outside ever because `<reason>`, but I do not really agree with the reasoning.

> It was designed with the idea that DockerRunner will always start a container for a test.

That's a design change of DockerRunner introduced with this PR, my comment is based on original design that is more permissive. The reason why I didn't design it like that with the initial build of the framework is because I can think of tests where we want to just `docker run` a single command and check the output, to check the exact, automated behavior of that (as was done with the version test), or, to run multiple docker containers to test an end-to-end use case, e.g. doing development on regtest, where there are scenarios where you need at least 2 nodes that connect to each other, for example when you want to work with getblocktemplate or auxpow RPC interfaces.